hcl/cty/object.go

```diff
@@ -0,0 +1,174 @@
+// Copyright 2022 Mineiros GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cty
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/mineiros-io/terramate/errors"
+	"github.com/mineiros-io/terramate/hcl"
+	"github.com/mineiros-io/terramate/project"
+	"github.com/zclconf/go-cty/cty"
+)
+
+// ErrCannotExtendObject is the error when an object cannot be extended.
+const ErrCannotExtendObject errors.Kind = "cannot extend object"
+
+type (
+	// Object is a object value supporting set at arbitrary paths using a
+	// dot notation.
+	//
+	// Eg.:
+	//   obj := cty.NewObject()
+	//   obj.Set("val", cty.NewObject())
+	//
+	// The snippet above creates the object below:
+	//   {
+	//       val = {}
+	//   }
+	//
+	// Then values can be set inside obj.val by doing:
+	//
+	//   obj.SetAt("val.test", 1)
+	Object struct {
+		// Keys is a map of key names to values.
+		Keys map[string]interface{}
+	}
+
+	// Value is a Hashicorp cty.Value wrapper.
+	Value struct {
+		Origin project.Path
+
+		val cty.Value
+	}
+
+	// DotPath represents a path inside the object using a dot-notation.
+	DotPath string
+)
+
+// NewObject creates a new object.
+func NewObject() *Object {
+	return &Object{
+		Keys: make(map[string]interface{}),
+	}
+}
+
+// Set a key value into object.
+func (obj *Object) Set(key string, value interface{}) {
+	if vvalue, ok := value.(Value); ok {
+		if vvalue.Raw().Type().IsObjectType() {
+			newobj := NewObject()
+			newobj.SetFrom(vvalue.Raw().AsValueMap())
+			value = newobj
+		}
+	}
+	obj.Keys[key] = value
+}
+
+// GetKeyPath retrieves the value at path.
+func (obj *Object) GetKeyPath(path DotPath) (interface{}, bool) {
+	parts := strings.Split(string(path), ".")
```

Comment:
Definitely not important, but here it seems to be the common use case for the new function [strings.Cut](https://pkg.go.dev/strings#Cut). It would help us avoid the re-joining.